pkg/types/symboltable/symboltable.go

```diff
@@ -0,0 +1,419 @@
+		return nil, err
+	}
+
+	var symbol gotypes.DataType
+	switch symbolType {
+	case gotypes.FunctionType:
+		var t gotypes.Function
+		if err := json.Unmarshal(b, &t); err != nil {
+			return nil, err
+		}
+		symbol = &t
+	case gotypes.MapType:
+		var t gotypes.Map
+		if err := json.Unmarshal(b, &t); err != nil {
+			return nil, err
+		}
+		symbol = &t
+	case gotypes.SliceType:
+		var t gotypes.Slice
+		if err := json.Unmarshal(b, &t); err != nil {
+			return nil, err
+		}
+		symbol = &t
+	case gotypes.StructFieldsItemType:
+		var t gotypes.StructFieldsItem
+		if err := json.Unmarshal(b, t); err != nil {
+			return nil, err
+		}
+		symbol = &t
+	case gotypes.StructType:
+		var t gotypes.Struct
+		if err := json.Unmarshal(b, &t); err != nil {
+			return nil, err
+		}
+		symbol = &t
+	case gotypes.SelectorType:
+		var t gotypes.Selector
+		if err := json.Unmarshal(b, &t); err != nil {
+			return nil, err
+		}
+		symbol = &t
+	case gotypes.InterfaceMethodsItemType:
+		var t gotypes.InterfaceMethodsItem
+		if err := json.Unmarshal(b, &t); err != nil {
+			return nil, err
+		}
+		symbol = &t
+	case gotypes.InterfaceType:
+		var t gotypes.Interface
+		if err := json.Unmarshal(b, &t); err != nil {
+			return nil, err
+		}
+		symbol = &t
+	case gotypes.EllipsisType:
+		var t gotypes.Ellipsis
+		if err := json.Unmarshal(b, &t); err != nil {
+			return nil, err
+		}
+		symbol = &t
+	case gotypes.ArrayType:
+		var t gotypes.Array
+		if err := json.Unmarshal(b, &t); err != nil {
+			return nil, err
+		}
+		symbol = &t
+	case gotypes.IdentifierType:
+		var t gotypes.Identifier
+		if err := json.Unmarshal(b, &t); err != nil {
+			return nil, err
+		}
+		symbol = &t
+	case gotypes.PointerType:
+		var t gotypes.Pointer
+		if err := json.Unmarshal(b, &t); err != nil {
+			return nil, err
+		}
+		symbol = &t
+	case gotypes.MethodType:
+		var t gotypes.Method
+		if err := json.Unmarshal(b, &t); err != nil {
+			return nil, err
+		}
+		//FIXME: methods and hash-names - see prefix
+		symbol = &t
+	case gotypes.ChannelType:
+		var t gotypes.Channel
+		if err := json.Unmarshal(b, &t); err != nil {
+			return nil, err
+		}
+		symbol = &t
+	default:
+		return nil, fmt.Errorf("JSON unmarshal: Unknown symbol type '%v'", symbolType)
+	}
+
+	return symbol, nil
+}
+
+func (o *SymbolDef) UnmarshalJSON(b []byte) error {
+	var objMap map[string]*json.RawMessage
+	var symbol gotypes.DataType
+	var err error
+
+	if err = json.Unmarshal(b, &objMap); err != nil {
+		return err
+	}
+
+	if err = json.Unmarshal(*objMap["name"], &o.Name); err != nil {
+		return err
+	}
+
+	if symbol, err = unmarshalDataType(*objMap["def"]); err != nil {
+		return err
+	}
+
+	o.Def = symbol
+	return nil
+}
+
+///////////////////////////////////////////////////////////////////
+
+type Package struct {
+	//TODO: really?
+	variables map[string]*SymbolDef
+	functions map[string]*SymbolDef
+	datatypes map[string]*SymbolDef
+	constants map[string]*SymbolDef
```

Comment:
There is no difference between variable and constant from a PoV of a symbol definition.